File: IT/Oggetti/Esercizi/C++/esercizio8.cpp
/* file esercizio8.cpp */

#include <iostream>

using std::cout;
using std::endl;
using std::string;
using std::ostream;

class Spaceman {
	private:
		string name;

	public:
		Spaceman(string);
		// necessaria la keyword const
		string getName() const;
		void setName(string);
		void floatInSpace();

	friend ostream& operator<<(ostream&, const Spaceman&);
};

Spaceman::Spaceman(string name)
{
	this->setName(name);
}

string Spaceman::getName() const {
	return this->name;
}

void Spaceman::setName(string name) {
	this->name = name;
}

ostream& operator<<(ostream& os, const Spaceman& s) {
	os << "Sono l'astronauta " << s.name << "!";
	return os;
}

void Spaceman::floatInSpace() {
	cout << this->getName() << " galleggia nello spazio..." << endl;
}

class Pilot : public Spaceman {
	public:
		Pilot(string name);

	friend ostream& operator<<(ostream&, const Pilot&);
};

class Explorer : public Spaceman {
	public:
		Explorer(string name);

	friend ostream& operator<<(ostream&, const Explorer&);
};

Pilot::Pilot(string name) : Spaceman(name) {}

ostream& operator<<(ostream &os, const Pilot &p) {
	os << "Sono il pilota " << p.getName() << ".";
	return os;
}

Explorer::Explorer(string name) : Spaceman(name) {}

ostream& operator<<(ostream &os, const Explorer &e) {
	os << "Sono l'esploratore " << e.getName() << ".";
	return os;
}

int main()
{
	Pilot buzz("Buzz Aldrin");
	Explorer neil("Neil Armstrong");

	cout << buzz << endl;
	cout << neil << endl;
}

File: IT/Oggetti/Esercizi/Java/Esercizio8.java
/* file Esercizio8.java */

class Spaceman {
	private String name;

	public Spaceman(String name) {
		setName(name);
	}

	public String getName() {
		return this.name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public void floatInSpace() {
		System.out.println(this.getName() + " si muove nello spazio...");
	}

	@Override
	public String toString() {
		return "Sono l'astronauta " + this.name + "!";
	}
}

class Pilot extends Spaceman {
	public Pilot(String name) {
		super(name);
	}

	@Override
	public String toString() {
		return "Sono il pilota " + this.getName();
	}
}

class Explorer extends Spaceman {
	public Explorer(String name) {
		super(name);
	}

	@Override
	public String toString() {
		return "Sono l'esploratore " + this.getName();
	}
}

public class Esercizio8 {
	public static void main(String [] args) {
		Pilot buzz = new Pilot("Buzz Aldrin");
		Explorer neil = new Explorer("Neil Armstrong");

		System.out.println(buzz);
		System.out.println(neil);
	}
}

